src/Components/Components/src/IHandleStateChange.cs

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// Interface implemented by components that receive notification of state changes.
```

Comment:
This feels confusing/wrong programming model wise. As far as I understand, state is internal to the component and the component is responsible for triggering renders when its state changes and it needs to rerender.

Reply:
I think I understand now what you are trying to accomplish here. I think this poses the issue that the component might rerender in many cases even if nothing changes, given that it doesn't have a good way to compare its previous state against the current state, and that might introduce performance penalties.

Reply:
The runtime needs to notify the component **somehow** that an event happened. This is the somehow - it's up to the component whether it wants to rerender or not. By default the answer is yes.

Reply:
I do agree with @javiercn, though @rynowak of course you're right that we need some way to signal that an event happened.

Perhaps we should change the name. Instead of `IHandleStateChange`, something like `IHandleAfterEvent`, since we're not really saying any state has changed, but just that some event has happened.

Actually I'm looking for all the places where we call `HandleStateChangeAsync` with some task other than `Task.CompletedTask`, and there's only one such place: `ComponentBase.cs` L171. That being the case, might it be possible to keep the earlier interface where we expect to be given `EventHandlerInvoker invoker` and `UIEventArgs args`, and then the recipient is the thing responsible for actually calling the `invoker`? If we could do that, it pretty much completely eliminates the risk that this interface gets misused as a way to invoke `StateHasChanged` on arbitrary components outside the intended lifecycle, preserving the encapsulation so that components know they only trigger that themselves (or the system does).

Perhaps there are other cases where we actually have to start with an existing `Task` rather than an `EventHandlerInvoker`. What do you think?

Reply:
Oh actually I guess there's always the scenario like in `SetValueHandler` where there's no delegate to pass, but rather we're just supplying notification that an event has already happened. So maybe you can ignore the second half of my suggestions there.

But I do still think `IHandleAfterEvent` might be a better name than `IHandleStateChange`.

Reply:
I'm going to change this to be void and called `IHandleAfterEvent` per feedback.

Reply:
I ended up changing this back to its original signature. The problem is that if we want the component to return a task, it's helpful for the component to be the one to invoke the delegate. Other solutions (unless we made it synchronous) become *really* complicated.